pallet/dex/src/lib.rs

```diff
@@ -0,0 +1,525 @@
+	traits::{
+		fungibles::{Balanced, Mutate},
+		tokens::Preservation,
+	},
+};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+type OrderOf<T> = Order<<T as frame_system::Config>::AccountId>;
+
+#[derive(Encode, Decode, Default, Eq, PartialEq, Clone, RuntimeDebug, TypeInfo)]
+pub struct TokenInfo {
+	pub asset_id: u32,
+	pub amount: u128,
+}
+
+#[derive(Encode, Decode, Eq, PartialEq, Clone, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub enum OrderType {
+	BUY,
+	SELL,
+}
+
+impl Default for OrderType {
+	fn default() -> Self {
+		OrderType::BUY
+	}
+}
+
+#[derive(Encode, Decode, Default, Eq, PartialEq, Clone, RuntimeDebug, TypeInfo)]
+pub struct Order<AccountId> {
+	counter: u64,       //order index
+	address: AccountId, //
+	pair: (u32, u32),   //AssetId_1 is base,  AssetId_2 is quote token
+	timestamp: u64,
+	order_type: OrderType,
+	amount_offered: u128,
+	amout_requested: u128,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::{ValueQuery, *},
+		Blake2_128Concat,
+	};
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig {
+		pub asset_ids: Vec<u32>,
+	}
+
+	impl Default for GenesisConfig {
+		fn default() -> Self {
+			GenesisConfig {
+				asset_ids: Default::default(),
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {
+			let bounded_token_infoes: BoundedVec<u32, ConstU32<{ u32::MAX }>> = self
+				.asset_ids
+				.clone()
+				.try_into()
+				.expect("genesis asset_ids are more than u32::MAX");
+
+			let mut index = 0;
+			self.asset_ids.iter().for_each(|asset_id| {
+				TokenIndex::<T>::insert(asset_id, index);
+				index += 1;
+			});
+
+			TokenInfoes::<T>::put(bounded_token_infoes);
+		}
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	/// The module configuration trait.
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Expose customizable associated type of asset transfer, lock and unlock
+		type Fungibles: Balanced<Self::AccountId>
+			+ Mutate<Self::AccountId, AssetId = u32, Balance = u128>;
+
+		type PrivilegedOrigin: EnsureOrigin<<Self as frame_system::Config>::RuntimeOrigin>;
+	}
+
+	/************* STORAGE ************ */
+	#[pallet::storage]
+	#[pallet::getter(fn user_token_infoes)]
+	pub type UserTokenInfoes<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId, //address
+		Blake2_128Concat,
+		u32, //asset id
+		TokenInfo,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn token_infoes)]
+	pub type TokenInfoes<T: Config> =
+		StorageValue<_, BoundedVec<u32, ConstU32<{ u32::MAX }>>, ValueQuery>;
```

Comment:
Does it work fine? I just a bit scared that it prealocates 4 bytes * u32::MAX. I always thought bounded vec reserves space and works more like array wrapper

Reply:
let me do some research.

Reply:
The size parameter is just a Phantom parameter.  See https://crates.parity.io/src/frame_support/storage/bounded_vec.rs.html#40